### PR TITLE
Add `hook_params` to SqlSensor.

### LINKED ISSUE #13750: Support Standard SQL in BigQuery Sensor
**Description**

A sql sensor which uses Standard SQL due to default one uses legacy sql

**Use case / motivation**

Currently (correct me if I am wrong!), the sql sensor only supports legacy sql. If I want to poke a BQ table, I do not think I can do that using standard sql right now.

**Are you willing to submit a PR?**

If community approves of this idea, sure!

### PR DESCRIPTION
**Pull request by kazanzhy**
This PR is based on: #17592 and closes: #13750
SqlSensor relies on underlying hooks that are automatically injected (DI) depending on the connection type provided to the SqlSensor. However, from time to time (but mainly in BigqueryHook) it is needed to pre-configure the "to-be-injected" hook because some default params do not fit the current config.
For example, if using SqlSensor with Bigquery it by default configures the SQL dialect to "legacy SQL", and there is no way to change this through configuration. Instead, the SqlSensor has to be extended and some functions are overwritten.

### CODE REVIEW

**Comment by kazanzhy** [comment]
Hi, @SamWheating.
Could you help me with this PR?

File: airflow/sensors/sql.py [hunk 1]
```diff
@@ -58,21 +61,22 @@ class SqlSensor(BaseSensorOperator):
...
         self.fail_on_empty = fail_on_empty
+        self.hook_params = hook_params
         super().__init__(**kwargs)
 
     def _get_hook(self):
         conn = BaseHook.get_connection(self.conn_id)
 
         allowed_conn_type = {
-            'google_cloud_platform',
+            'gcpbigquery',
```
Comment (kazanzhy) [inline airflow/sensors/sql.py]: As I understand, to use some connection in this sensor it hook should be an instance of DbApiHook.
But GoogleBaseHook which returns for the google_cloud_platform isn't that one.
Am I correct?

Comment (SamWheating) [reply airflow/sensors/sql.py]: I think that if this were the case then it might be easier to instantiate the hook and then check if its a subclass of `dbApiHook` using something like 

```python
conn = BaseHook.get_connection(self.conn_id)
hook = conn.get_hook(self.hook_params)
if issubclass(hook, DbApiHook):
 ....
```

So maybe this is isn't supported for all DbApiHooks? Let me see if the original PR which added this offered any explanation.

Comment (SamWheating) [reply airflow/sensors/sql.py]: Yeah, this certainly isn't a list of all DbApiHooks, since its missing `druid` and `eleasticsearch` amongst others.

Comment (SamWheating) [reply airflow/sensors/sql.py]: I found the PR which introduced this change:
https://github.com/apache/airflow/pull/4723/files

And it looks like at the time at the time, the `get_hook` function looked very different and returned a BigQueryHook for the `google_cloud_platform` conn type.:
https://github.com/apache/airflow/blob/b1acc5508a4b9aaa7c98104ba31ab0623c95f4aa/airflow/models/connection.py#L184-L190

So I think you're right that this is incorrect, since this would allow someone to create a SqlSensor with a `ComputeEngineSSHHook`, which certainly wouldn't work - nice catch!

cc @XD-DENG I think a refactor might be required here - thoughts on either:
 - updating the list of conn_types to include some new connections
 - changing the check here to assert that a connection is a subclass of a dbApiHook (and thus has a get_records method).

File: airflow/sensors/sql.py [hunk 2]
```diff
@@ -90,7 +94,7 @@ def _get_hook(self):
                 f"Connection type ({conn.conn_type}) is not supported by SqlSensor. "
                 + f"Supported connection types: {list(allowed_conn_type)}"
             )
-        return conn.get_hook()
+        return conn.get_hook(self.hook_params)
```
Comment (SamWheating) [inline airflow/sensors/sql.py]:
```suggestion
        return conn.get_hook(hook_params=self.hook_params)
```

Using a keyword argument might make this more readable and less likely to break if someone later adds another argument to the `get_hook` function.

However this might be more personal opinion, thoughts?

Comment (kazanzhy) [reply airflow/sensors/sql.py]: Definitely agree with you

File: airflow/models/connection.py [hunk 3]
```diff
@@ -289,8 +289,11 @@ def rotate_fernet_key(self):
         if self._extra and self.is_extra_encrypted:
             self._extra = fernet.rotate(self._extra.encode('utf-8')).decode()
 
-    def get_hook(self):
-        """Return hook based on conn_type."""
+    def get_hook(self, hook_params: Dict = {}):
```
Comment (dinigo) [inline airflow/models/connection.py]: It is discouraged to pass a mutable object as a default. See this SO https://stackoverflow.com/q/1132941/621058 . Instead you can null it by default and if it's null assign it later

```python
def get_hook(self, hook_params: Dict = None):
  if hook_params is None:
    hook_params = {}
  # ...
```

**Comment by dinigo** [comment]
Hello there. Been on holiday for some time, thats why #17592 was stalled. I see you get good progress here, including adding the missing DbHooks. Tell me If I can contribute somehow, just don't want to duplicate efforts

File: airflow/models/connection.py (hunk 3 above)
Comment (kazanzhy) [reply airflow/models/connection.py]: Yeah, I've also had to deal with the problem described in SO.
But in this case, Dict is just a container and hasn't been updated in the method.

Could you please write an example in the context of SqlSensor when we can receive the behavior similar to SO?
Because now, in my opinion, it looks like code sophistication.

**Comment by eladkal** [comment]
hi @kazanzhy recently https://github.com/apache/airflow/pull/18718 was merged which is similar.
I see that you are depended on https://github.com/apache/airflow/pull/17592 but since it's stale maybe you want to take over and integrate the needed changes to this PR?
This should be a relatively small change probably there is no real need to split the change into two PRs.

**Comment by kazanzhy** [comment]
Hi @eladkal.
As you see this PR was before #18718 and I didn't see that it's referred to this PR.
I'll make rebase to the current main branch and update this PR.
Thank you for your comment

File: airflow/models/connection.py (hunk 3 above)
Comment (dstandish) [reply airflow/models/connection.py]: I understand your point @kazanzhy but I agree we should just stick with the practice of using `None` as the default for `Optional[Dict]` that is, as far as I know, uniformly adhered to in this repo.

If you leave it, people will always give it a second look when they see it.  And pycharm complains about it.  And using `None` for an optional arg is also a common pattern that I think is good aside from the mutability issue, just sends the message, you don't need to pass this, and we don't forward anything in this case.

File: airflow/sensors/sql.py [hunk 4]
```diff
@@ -58,21 +61,34 @@ class SqlSensor(BaseSensorOperator):
...
-        self, *, conn_id, sql, parameters=None, success=None, failure=None, fail_on_empty=False, **kwargs
+        self, 
+        *, 
+        conn_id, 
+        sql, 
+        parameters=None, 
+        success=None, 
+        failure=None, 
+        fail_on_empty=False, 
+        hook_params={},
```
Comment (dstandish) [inline airflow/sensors/sql.py]: same as above

**Review by dstandish — CHANGES_REQUESTED** [review]
I think this needs to be split into 2 prs. This one should only add hook_params.  Changing the conn types should be done separately.

File: airflow/sensors/sql.py (hunk 1 above)
Comment (eladkal) [reply airflow/sensors/sql.py]: We actually check for subclass of `DbApiHook` in:
https://github.com/apache/airflow/blob/dd410fd3c9de14e94034dcb4ccae52bbf5216199/airflow/operators/sql.py#L68-L73

So we should do here the same.
I'm not sure why you mention druid and elasticsearch as exceptions?
https://github.com/apache/airflow/blob/dd410fd3c9de14e94034dcb4ccae52bbf5216199/airflow/providers/apache/druid/hooks/druid.py#L131

https://github.com/apache/airflow/blob/dd410fd3c9de14e94034dcb4ccae52bbf5216199/airflow/providers/elasticsearch/hooks/elasticsearch.py#L27

Comment (kazanzhy) [reply airflow/sensors/sql.py]: Done

**Comment by kazanzhy** [comment]
Hi @eladkal @dstandish. 
Please take a look at the latest changes.
I've used changes from PR #18718 and doesn't make changes in `airflow/models/connection.py`.
Also, I'm checking the hook if it is an instance of DbApiHook.

File: airflow/sensors/sql.py [hunk 5]
```diff
@@ -58,39 +62,35 @@ class SqlSensor(BaseSensorOperator):
...
-            'postgres',
-            'presto',
-            'snowflake',
-            'sqlite',
-            'trino',
-            'vertica',
-        }
-        if conn.conn_type not in allowed_conn_type:
+        hook = conn.get_hook(hook_kwargs=self.hook_kwargs)
+        if not isinstance(hook, DbApiHook):
```
Comment (dstandish) [inline airflow/sensors/sql.py]: It's not clear from the PR description why you're making the change with `allowed_conn_types`.  It doesn't seem like it's necessary as part of adding `hook_kwargs`, but seems completely unrelated.

Probably it makes sense to be in a separate PR.  But at least the description should be updated to explain what the reason is for this change.

**Comment by kazanzhy** [comment]
I updated the description. Also changed name of the parameter to `hook_params` which is similar to BaseSqlOperator.
I think the parameter of the method `Connection.get_hook()` also might be changed to `hook_params`

File: airflow/sensors/sql.py [hunk 6]
```diff
@@ -58,39 +62,35 @@ class SqlSensor(BaseSensorOperator):
...
-        self, *, conn_id, sql, parameters=None, success=None, failure=None, fail_on_empty=False, **kwargs
+        self,
+        *,
+        conn_id,
+        sql,
+        parameters=None,
+        success=None,
+        failure=None,
+        fail_on_empty=False,
+        hook_params=None,
```
Comment (uranusjr) [inline airflow/sensors/sql.py:72]: I wonder if we should use `hook_kwargs` for consistency with `Connection.get_hook()`.

File: airflow/sensors/sql.py [hunk 7]
```diff
@@ -58,39 +62,35 @@ class SqlSensor(BaseSensorOperator):
...
+        **kwargs
     ):
         self.conn_id = conn_id
         self.sql = sql
         self.parameters = parameters
         self.success = success
         self.failure = failure
         self.fail_on_empty = fail_on_empty
+        self.hook_params = {} if hook_params is None else hook_params
```
Comment (uranusjr) [inline airflow/sensors/sql.py]: The `if-else` is not needed since `Connection.get_hook()` handled None automatically.

```suggestion
        self.hook_params = hook_params
```

File: airflow/sensors/sql.py (hunk 5 above)
Comment (uranusjr) [reply airflow/sensors/sql.py]: Same feeling, this should be a separate PR.

File: airflow/sensors/sql.py (hunk 6 above)
Comment (kazanzhy) [reply airflow/sensors/sql.py:72]: I definitely agree with you.
But this inconsistency was provided by this PR (https://github.com/apache/airflow/pull/18718), where BaseSqlOperator takes `hook_params` and Connection takes `hook_kwargs`.
If this PR will be merged I'll create another one that resolves it.

Should I create an issue for that?
Resolve conflicting parameter SqlSensor(hook_params), BaseSqlOperator(hook_params), Connection(_**hook_kwargs**_)

Comment (uranusjr) [reply airflow/sensors/sql.py:72]: Yes let’s open a new issue and fix this separately. Thankfully we caught this rather quickly so we don’t need to worry about backwards compatibility as long as we can amend it before 2.3 is released!

Comment (kazanzhy) [reply airflow/sensors/sql.py:72]: Done
https://github.com/apache/airflow/issues/19638

File: airflow/sensors/sql.py (hunk 5 above)
Comment (kazanzhy) [reply airflow/sensors/sql.py]: Done.
https://github.com/apache/airflow/pull/19639

**Review by uranusjr — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

**Review by dstandish — APPROVED** [review]

**Review by kaxil — APPROVED** [review]

**Review by potiuk — APPROVED** [review]
